From 1f86ef1fa588430b1296760affeb770920b8c3f5 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Tue, 22 Aug 2023 01:33:03 -0700
Subject: [PATCH 54/78] UBUNTU: SAUCE: apparmor4.0.0 [57/82]: prompt - lock
 down prompt interface

BugLink: https://bugs.launchpad.net/bugs/2028253

While the prompt interface is supposed to be open and permissions
handled via namespacing. There are no current users that require
it be open and there are some namespacing issues that need to
be resolved. So Lock the interface down until the namespacing
issues are all fixed.

Signed-off-by: John Johansen <john.johansen@canonical.com>
(cherry picked from https://gitlab.com/jjohansen/apparmor-kernel)
Signed-off-by: Andrea Righi <andrea.righi@canonical.com>
---
 security/apparmor/apparmorfs.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/security/apparmor/apparmorfs.c b/security/apparmor/apparmorfs.c
index b32931b67024..bf0a7180f8a5 100644
--- a/security/apparmor/apparmorfs.c
+++ b/security/apparmor/apparmorfs.c
@@ -616,6 +616,8 @@ static int listener_release(struct inode *inode, struct file *file)
 {
 	struct aa_listener *listener = file->private_data;
 
+	if (!aa_current_policy_admin_capable(NULL))
+		return -EPERM;
 	if (listener)
 		aa_put_listener(listener);
 
@@ -626,6 +628,8 @@ static int listener_open(struct inode *inode, struct file *file)
 {
 	struct aa_listener *listener;
 
+	if (!aa_current_policy_admin_capable(NULL))
+		return -EPERM;
 	listener = aa_new_listener(NULL, GFP_KERNEL);
 	if (!listener)
 		return -ENOMEM;
@@ -726,6 +730,8 @@ static long listener_ioctl(struct file *file, unsigned int cmd,
 {
 	struct aa_listener *listener = file->private_data;
 
+	if (!aa_current_policy_admin_capable(NULL))
+		return -EPERM;
 	if (!listener)
 		return -EINVAL;
 
@@ -752,6 +758,9 @@ static __poll_t listener_poll(struct file *file, poll_table *pt)
 	struct aa_listener *listener = file->private_data;
 	__poll_t mask = 0;
 
+	if (!aa_current_policy_admin_capable(NULL))
+		return -EPERM;
+
 	if (listener) {
 		spin_lock(&listener->lock);
 		poll_wait(file, &listener->wait, pt);
-- 
2.43.0

